packages/react/src/components/ContentSwitcher/ContentSwitcher.js

```diff
@@ -5,125 +5,124 @@
-    children: PropTypes.node,
-
-    /**
-     * Specify an optional className to be added to the container node
-     */
-    className: PropTypes.string,
-
-    /**
-     * Specify an `onChange` handler that is called whenever the ContentSwitcher
-     * changes which item is selected
-     */
-    onChange: PropTypes.func.isRequired,
-
-    /**
-     * Specify a selected index for the initially selected content
-     */
-    selectedIndex: PropTypes.number,
-  };
-
-  static defaultProps = {
-    selectedIndex: 0,
-  };
-
-  static getDerivedStateFromProps({ selectedIndex }, state) {
-    const { prevSelectedIndex } = state;
-    return prevSelectedIndex === selectedIndex
-      ? null
-      : {
-          selectedIndex,
-          prevSelectedIndex: selectedIndex,
-        };
+function ContentSwitcher({
+  children,
+  className: customClassName,
+  onChange,
+  selectedIndex: controlledSelectedIndex = 0,
+  ...rest
+}) {
+  const switchRefs = [];
+  const className = cx(`${prefix}--content-switcher`, customClassName);
+  const isMounted = useRef(false);
+  const savedOnChange = useRef(null);
+  const [selectedIndex, setSelectedIndex] = useState(controlledSelectedIndex);
+  const [prevControlledIndex, setPrevControlledIndex] = useState(
+    controlledSelectedIndex
+  );
+
+  if (controlledSelectedIndex !== prevControlledIndex) {
+    setSelectedIndex(controlledSelectedIndex);
+    setPrevControlledIndex(controlledSelectedIndex);
   }
 
-  handleItemRef = index => ref => {
-    this._switchRefs[index] = ref;
-  };
-
-  handleChildChange = data => {
-    // the currently selected child index
-    const { selectedIndex } = this.state;
-    // the newly selected child index
-    const { index } = data;
-    const { key } = data;
-
-    if (matches(data, [keys.ArrowRight, keys.ArrowLeft])) {
-      const nextIndex = getNextIndex(
-        key,
-        selectedIndex,
-        this.props.children.length
-      );
-      this.setState(
-        {
-          selectedIndex: nextIndex,
-        },
-        () => {
-          const switchRef = this._switchRefs[nextIndex];
-          switchRef && switchRef.focus();
-          this.props.onChange(data);
-        }
-      );
-    } else {
-      if (selectedIndex !== index) {
-        this.setState({ selectedIndex: index }, () => {
-          const switchRef = this._switchRefs[index];
-          switchRef && switchRef.focus();
-          this.props.onChange(data);
-        });
+  // Always keep track of the latest `onChange` prop to use in our focus effect
+  // handler
+  useEffect(() => {
+    savedOnChange.current = onChange;
+  }, [onChange]);
+
+  // Only fire this effect after the first render, otherwise we end up
+  // autofocusing the selected index when this component is rendered
+  useEffect(() => {
+    if (!isMounted.current) {
+      return;
+    }
+
+    const ref = switchRefs[selectedIndex];
+    if (ref && document.activeElement !== ref.current) {
+      ref.focus && ref.focus();
```

Comment:
Focus change has to be trigged by user gesture. Please see: https://github.com/carbon-design-system/carbon/pull/2852#discussion_r290605855

Reply:
@asudoh the way that `selectedIndex` changes is through a user gesture like `onClick` or `onKeyDown` 👍 When that happens, this effect is run.

Reply:
Just noticed a bug with the controlled selected index causing this focus and will totally fix that if that's what you were talking about!

Reply:
Yes this is what I was talking about. Thanks!